service/src/main/java/bio/terra/workspace/service/workspace/flight/delete/cloudcontext/DeleteResourcesStep.java

```diff
@@ -0,0 +1,143 @@
+  public StepResult doStep(FlightContext context) throws InterruptedException, RetryException {
+    // Jackson does not deserialize the list properly with a TypeReference. The unchecked for works.
+    ResourceDeleteFlightPair[] resourcePairs =
+        FlightUtils.getRequired(
+            context.getWorkingMap(),
+            RESOURCE_DELETE_FLIGHT_PAIR_LIST,
+            ResourceDeleteFlightPair[].class);
+
+    Stairway stairway = context.getStairway();
+    // Loop through the resources, deleting them one at a time
+    for (ResourceDeleteFlightPair pair : resourcePairs) {
+      try {
+        FlightState flightState = stairway.getFlightState(pair.flightId());
+        FlightStatus flightStatus = flightState.getFlightStatus();
+
+        // If we already finished this flight, we continue through the list of resources to delete
+        if (flightStatus == SUCCESS) {
+          continue;
+        }
+
+        // If the resource delete flight completed with an error. We complete this delete
+        // flight with a different error, reporting the failure
+        if (flightStatus == ERROR || flightStatus == FATAL) {
+          handleResourceFlightFailure(flightState, pair.resourceId());
+        }
+
+        // The flight exists and is still running in Stairway. We wait...
+        waitForFlight(pair, stairway);
+      } catch (FlightNotFoundException e) {
+        launchFlightAndWait(pair, stairway);
+      }
+    }
+
+    return StepResult.getStepResultSuccess();
+  }
+
+  private void handleResourceFlightFailure(FlightState flightState, UUID resourceId) {
+    String message;
+    Throwable throwable;
+    Optional<Exception> flightException = flightState.getException();
+    if (flightException.isPresent()) {
+      throwable = flightException.get();
+      message = throwable.getMessage();
+    } else {
+      throwable = null;
+      message = "flight failed with no error message";
+    }
+    throw new CloudContextDeleteException(
+        String.format(
+            "Cloud context delete failed because resource %s could not be deleted. Cause was: %s",
+            resourceId, message),
+        throwable);
+  }
+
+  private void launchFlightAndWait(ResourceDeleteFlightPair pair, Stairway stairway)
+      throws InterruptedException {
+    cloudContextService.launchDeleteFlight(
+        controlledResourceService, workspaceUuid, pair.resourceId(), pair.flightId(), userRequest);
+    waitForFlight(pair, stairway);
+  }
+
+  // We are VERY patient with deletion flights. GCP buckets can take many hours to delete if
+  // they hold a lot of data.
+  private void waitForFlight(ResourceDeleteFlightPair pair, Stairway stairway)
+      throws InterruptedException {
+    FlightState flightState;
+    try {
+      flightState =
+          RetryUtils.getWithRetry(
+              FlightUtils::flightComplete,
+              () -> stairway.getFlightState(pair.flightId()),
+              Duration.ofHours(7), /* total duration */
```

Comment:
Good to know. It's a little surprising it can take that long, but that's outside of our control.